_adr/83/index.adoc

```diff
@@ -0,0 +1,155 @@
+.. Having knowledge of the AMS endpoints and how to deal with them
+.. Having knowledge of the AMS terms used for designing the instance types (RHOSAKTrial for eval/developer and RHOSAK for standard)
+.. Enquire the quota assigned using the AMS endpoints, loop through the resources and check that enough quota is still available for product RHOSAK and resource RHOSAK
+.. If you have no quota assigned for STANDARD instances (RHOSAK), fallback to EVAL
+.. If you have quota for STANDARD (RHOSAK), but it is 0, fallback to EVAL (RHOSAKTrial)
+.. If you have quota for STANDARD (RHOSAK) and it is greater than 0 but all the granted quota is consumed, the creation should fail
+. Verify that at least one OCD Cluster has capacity for the instance you are going to create
+
+Another point of attention is that currently, any change in the flow requires an update to all the actors/consumers who are carrying out those checks.
+
+## Goals
+* API consumers can query a new KAS Fleet Manager Permission API to check if they can perform a certain action and receive a response with the outcome (allowed/denied) containing all the granted resources or, eventually, the missing requirements.
+
+## Non-goals
+
+## Current architecture
+* Kas Fleet Manager currently only allows the ‘try and see if it works’ approach: if what you are trying to do fails, Kas Fleet Manager gives you some details on went wrong, but you have no way of checking in advance if what you want to do is possible
+* API Consumers (like UI) are doing the required checkss into their codebase and every time a change occurs, both Kas Fleet Manager and UI have to be updated to reflect the change
+
+## Proposed architecture
+* We need to create a new API.
+The API will be authenticated and will allow the caller to check in advance whether he can perform a certain operation.
+A different endpoint will be provided for each verb-resource pair, wich, in turn, will provide a schema defining exactly what the answer will look like.
++
+```
+GET /api/v1/can/{verb}/{resource}?deep=true
+```
++
+The API will stop at the first error unless `deep=true` is specified
+Output of the API will be a JSON with at least the following content:
++
+```
+{
+  allowed: boolean,
+  reasons: [{
+    code: string,
+    description: string,
+    metadata: {             // additional details
+      {key string}: {value string}
+    }
+  }]
+}
+```
++
+The rest of the content will depend on the action-resource pair.
++
+Some real world example for an invocation of the `/can/create/kafka` could be:
+
+.. Just one quota available: `standard.x2` won't be returned
++
+```
+{
+  allowed: true,
+  plans: ['standard.x1'],
+  quota: {
+    granted: 6,
+    used: 5
+  }
+}
+```
+
+.. More than one quota available:
++
+```
+{
+  allowed: true,
+  plans: ['standard.x1', 'standard.x2'],
+  quota: {
+    granted: 6,
+    used: 5
+  }
+}
+```
+
+.. Terms not accepted
++
+```
+{
+  allowed: false,
+  reasons: [{
+    code: "ERR-TERMS-01",
```

Comment:
Will it be possible to use the same error codes pattern (see the response output of `/api/kafkas_mgmt/v1/errors` and observe the error code) as the one we current have in kas-fleet-manager? Introducing another form of error code might requires clients having to reason about different error codes patterns which can be inconsistent

Reply:
Sure.